Make pkgcheck happy

blackteahamburger opened the pull request:
I found a lot of imperfections in the ebuilds in this repository, and here are some fixes (partial).

In addition, some ebuilds that are too old and already in the main tree are dropped.

blackteahamburger commented:
Next step is switching chromium-based ebuilds to llvm-r1. I think we can drop gcc support.

PF4Public commented:
> I found a lot of imperfections in the ebuilds in this repository

:D Like I mentioned in another issue here, I value maintainability more than perfection. Especially that I have less spare time lately.

>here are some fixes

Thanks for the fixes though

> already in the main tree are dropped

That is a good thing.

>www-client/opera: tree clean

Sorry, but no, it's personal

> add .gitignore

This file does not belong here

>net-im/element-desktop: make pkgcheck happy, drop 1.11.{86..89}

But it would make me unhappy because `IUSE` go all the way down under the first screen and I'll need to scroll every time I need to add electron slot. Additionally this will break my auto-update script because you've removed `SRC_URI=""`, which would make me unhappy even more.

> dev-util/electron/files/node-gyp 100755 → 100644

Did you intentionally change the permissions?

Ungoogled, cromite and electron follow the same structure for ease of diffing against each other and mainline Gentoo.

>I think we can drop gcc support.

It almost never worked, but was left for anyone, who is willing to try and fix.

> And I want to add mold support.

Will it work with PGO and LTO?

blackteahamburger commented:
> > www-client/opera: tree clean
> 
> Sorry, but no, it's personal

OK, then revert it.
> > add .gitignore
> 
> This file does not belong here

For convenience for pkgcheck.
> > net-im/element-desktop: make pkgcheck happy, drop 1.11.{86..89}
> 
> But it would make me unhappy because `IUSE` goes all the way down under the first screen and I'll have to scroll every time I need to add electron slot. Additionally this will break my auto-update script because you've removed `SRC_URI=""`, which would make me unhappy even more. The former could be automated too, but I just don't have time for that right now as well as for rewriting and debugging auto-update.

OK, then revert it.
> > dev-util/electron/files/node-gyp 100755 → 100644
> 
> Did you intentionally change the permissions?

What pkgcheck suggests.
> Ungoogled, cromite and electron follow the same structure for ease of diffing against each other and mainline Gentoo.
> 
> > I think we can drop gcc support.
> 
> It almost never worked, but was left for anyone, who is willing to try and fix

But it makes switching to llvm-r1 inconvenient.
> > And I want to add mold support.
> 
> Will it work with PGO and LTO?

Testing.

PF4Public commented:
> > > add .gitignore
> > 
> > 
> > This file does not belong here
> 
> For convenience for pkgcheck. Overlays I know all have it.

you can add .gitignore to .gitignore :) I have other excludes there, vscode for example

> > > dev-util/electron/files/node-gyp 100755 → 100644
> > 
> > 
> > Did you intentionally change the permissions?
> 
> What pkgcheck suggests.

then it is fine

> > Ungoogled, cromite and electron follow the same structure for ease of diffing against each other and mainline Gentoo.
> > > I think we can drop gcc support.
> > 
> > 
> > It almost never worked, but was left for anyone, who is willing to try and fix
> 
> But it makes switching to llvm-r1 inconvenient. And there is makes no sense to retain gcc support since lto & pgo is unavailable with gcc. And as chromium updates, making gcc support work will become more and more difficult.

were you referring to the structure or gcc support. gcc support is not so important as the structure to compare against mainlain chromium, although if you don't do lots of changes, that might still be fine

blackteahamburger commented:
As you requested, I re-added gcc support.

> > > > add .gitignore
> > > 
> > > 
> > > This file does not belong here
> > 
> > 
> > For convenience for pkgcheck. Overlays I know all have it.
> 
> you can add .gitignore to .gitignore :) I have other excludes there, vscode for example
> 

Then why don't we add all excludes to it?
> > > Ungoogled, cromite and electron follow the same structure for ease of diffing against each other and mainline Gentoo.
> > > > I think we can drop gcc support.
> > > 
> > > 
> > > It almost never worked, but was left for anyone, who is willing to try and fix
> > 
> > 
> > But it makes switching to llvm-r1 inconvenient. And there is makes no sense to retain gcc support since lto & pgo is unavailable with gcc. And as chromium updates, making gcc support work will become more and more difficult.
> 
> were you referring to the structure or gcc support. gcc support is not so important as the structure to compare against mainline chromium, although if you don't do lots of changes, that might still be fine

OK, I re-added gcc support.
> > > And I want to add mold support.
> > 
> > 
> > Will it work with PGO and LTO?
> 
> Testing.
Should work, but I my memory ran out in the final link. I have added a warning of OOM.

blackteahamburger commented:
The differences between the original ebuild and the ebuild in the main tree may be intentional, but I don't know. If so, please feel free to point it out.

Here are some explanations of the changes:

- Mask system-abseil-cpp use flag: avoid the need for "enable system-abseil-cpp -> failed -> disable system-abseil-cpp" process for newcomers, make the failure more explicit.
- Download chromium-patches from upstream: make it unnecessary to manually download patch, keep the patch up to date, and avoid missing patches. The current patch is out of date, and with the latest patch, `chromium-128-cfi-split-lto-unit.patch` is no longer needed.
- Drop thinlto use flag: sync with main tree, making enabling of lto more transparent for users who set system-wide lto.
- Drop optimize-thinlto use flag: follow main tree & chromium upstream default since it is impossible to make it a default option while maintaining the ability to switch now. It does not affect too much since it is only used on important targets.
- Drop qt5: follow main tree. Qt5 is being removed now.
- Get rid of system-ffmpeg build faliure: I created gentoo/gentoo#40469 to get rid of system-ffmpeg build failure once and for all. Added ffmpeg ebuilds with av-stream-get-first-dts use flag for transition before the pr is merged.
- Remove import_instr_limit change: default value is 30 now
- Remove `-Wl,--thinlto-jobs` set: avoid overriding flag set by user

PF4Public commented:
> > > > > add .gitignore
> > > > 
> > > > 
> > > > This file does not belong here
> > > 
> > > 
> > > For convenience for pkgcheck. Overlays I know all have it.
> > 
> > 
> > you can add .gitignore to .gitignore :) I have other excludes there, vscode for example
> 
> Then why don't we add all excludes to it?

Because in a month comes someone with intellij, then someone with emacs or whatever. Everyone just keeps this file to themselves and we're good.

> The differences between the original ebuild and the ebuild in the main tree may be intentional, but I don't know. If so, please feel free to point it out.

Do you have something specific on your mind? Presently the only differences I can name off the top of my head are: ungoogled-chromium from my overlay builds against libstdc++ and avoids rust (which apparently starts causing issues).

> * Mask system-abseil-cpp use flag

I'm not masking anything. If it's not on by default — it may fail. Additionally I'm not willing to unmask it if I'm up to try and fix it someday later. Maybe a little bit inconvenient for an end-user, but my apologies, I would favour convenience for myself.

> * Download chromium-patches from upstream: make it unnecessary to manually download patch, keep the patch up to date, and avoid missing patches.

I remember I separated the patch at the point where they stopped removing rust. Additionally sometimes they didn't have a patch at the time I was preparing an update, so I had to update it myself anyway.

> * Drop thinlto use flag: sync with main tree, making enabling of lto more transparent for users who set system-wide lto.

Well, I considered that myself, but then you'll ideally have to provide lto for gcc, so I left it as is.

> * Get rid of system-ffmpeg build faliure: I created [media-video/ffmpeg: add av-stream-get-first-dts use flag gentoo/gentoo#40469](https://github.com/gentoo/gentoo/pull/40469) to get rid of system-ffmpeg build failure once and for all. Added ffmpeg ebuilds with av-stream-get-first-dts use flag for transition before the pr is merged.

I'm a little bit pessimistic about your PR, but we'll see how it does. I don't see any benefits adding ffmpeg here. I don't have time to maintain those ebuilds.

> `system-png? ( media-libs/libpng:=[-apng(-)] )`

I remember someone told me something along the lines: _"they know nothing and break thing with this -apng, get rid of it"_. So I don't know now :)

> Drop qt5: follow main tree. Qt5 is being removed now.
> Remove import_instr_limit change: default value is 30 now

Thanks!

> `debug? ( !official )`
> `myconf_gn+=" is_debug=false"`

You somehow lost the conditional

> `[[ -z "${NODIE}" ]] && die "Please update the ebuild."`

Oh that's nice. Have you tested that it works how you intended?

> `append-cppflags -U_GLIBCXX_ASSERTIONS #See #318`

Fine, fine, you win :)

Why did you remove `icall` and `cast`?

> `if use x86; then` 

[Excuse me](https://github.com/PF4Public/ungoogled-chromium-bin/releases/tag/132.0.6834.159#:~:text=i686.tar.bz2,last%20week)? 

> Remove -Wl,--thinlto-jobs set: avoid overriding flag set by user

What would it be set to if user has nothing there?

> myconf_gn+=" devtools_skip_typecheck=false"

You shouldn't remove that. It solves an issue with `-official`.

> `ulimit -n 1048576`

BTW, does it really work with unchanged limits.conf? This line would be just ignored, would it not?



I should've better commented my code 🤦🏻

blackteahamburger commented:
> > > > > > add .gitignore
> > > > > 
> > > > > 
> > > > > This file does not belong here
> > > > 
> > > > 
> > > > For convenience for pkgcheck. Overlays I know all have it.
> > > 
> > > 
> > > you can add .gitignore to .gitignore :) I have other excludes there, vscode for example
> > 
> > 
> > Then why don't we add all excludes to it?
> 
> Because in a month comes someone with intellij, then someone with emacs or whatever. Everyone just keeps this file to themselves and we're good.

Got it.
> 
> > The differences between the original ebuild and the ebuild in the main tree may be intentional, but I don't know. If so, please feel free to point it out.
> 
> Do you have something specific on your mind? Presently the only differences I can name off the top of my head are: ungoogled-chromium from my overlay builds against libstdc++ and avoids rust (which apparently starts causing issues).

The differences I say are just the slight ones like what you point out below.
> > * Mask system-abseil-cpp use flag
> 
> I'm not masking anything. If it's not on by default — it may fail. Additionally I'm not willing to unmask it if I'm up to try and fix it someday later. Maybe a little bit inconvenient for an end-user, but my apologies, I would favour convenience for myself.

Got it.
> > * Download chromium-patches from upstream: make it unnecessary to manually download patch, keep the patch up to date, and avoid missing patches.
> 
> I remember I separated the patch at the point where they stopped removing rust. Additionally sometimes they didn't have a patch at the time I was preparing an update, so I had to update it myself anyway.

Got it. Then I would substitute the patch in filesdir and add `chromium-117-material-color-include.patch` (still needed?).
> > * Drop thinlto use flag: sync with main tree, making enabling of lto more transparent for users who set system-wide lto.
> 
> Well, I considered that myself, but then you'll ideally have to provide lto for gcc, so I left it as is.

There aren’t that many ideals... filter-lto for gcc is seen in some ebuilds.
> > * Get rid of system-ffmpeg build faliure: I created [media-video/ffmpeg: add av-stream-get-first-dts use flag gentoo/gentoo#40469](https://github.com/gentoo/gentoo/pull/40469) to get rid of system-ffmpeg build failure once and for all. Added ffmpeg ebuilds with av-stream-get-first-dts use flag for transition before the pr is merged.
> 
> I'm a little bit pessimistic about your PR, but we'll see how it does. I don't see any benefits adding ffmpeg here. I don't have time to maintain those ebuilds.
> 
> > `system-png? ( media-libs/libpng:=[-apng(-)] )`
> 
> I remember someone told me something along the lines: _"they know nothing and break thing with this -apng, get rid of it"_. So I don't know now :)

The just follow main tree and wait for a real issue report.
> > `debug? ( !official )`
> > `myconf_gn+=" is_debug=false"`
> 
> You somehow lost the conditional

You'd better say the main tree somehow lost the conditional. I don't know the reason but just consider that it was there but was deleted by main tree sometime.
> > `[[ -z "${NODIE}" ]] && die "Please update the ebuild."`
> 
> Oh that's nice. Have you tested that it works how you intended?

Again, following main tree, added NODIE check.
> Why did you remove `icall` and `cast`?

As above, unclear.
> > `if use x86; then`
> 
> [Excuse me](https://github.com/PF4Public/ungoogled-chromium-bin/releases/tag/132.0.6834.159#:~:text=i686.tar.bz2,last%20week)?

What do you mean? It now just do `filter-flags "-g*"` all the time. x86 support is well retained.
> > Remove -Wl,--thinlto-jobs set: avoid overriding flag set by user
> 
> What would it be set to if user has nothing there?

Uses all hyper-threads. See https://clang.llvm.org/docs/ThinLTO.html#controlling-backend-parallelism.
> > myconf_gn+=" devtools_skip_typecheck=false"
> 
> You shouldn't remove that. It solves an issue with `-official`.

Thanks, it's #142. Should add a comment.
> > `ulimit -n 1048576`
> 
> BTW, does it really work with unchanged limits.conf? This line would be just ignored, would it not?
> 
> I should've better commented my code 🤦🏻

The only possibility is that it has effect, see https://github.com/rui314/mold/issues/336. I have searched for the reason for quite a long time before...

PF4Public commented:
> add `chromium-117-material-color-include.patch` (still needed?).
> 
> > >

I don't think it is needed anymore: https://source.chromium.org/chromium/chromium/src/+/main:third_party/material_color_utilities/src/cpp/palettes/tones.cc;drc=13434b50dcb64a482cc91191f8cf6151d90f5465;l=19

> I don't know the reason but just consider that it was there but was deleted by main tree sometime

I restored this flag intentionally

> I wonder why they did nothing for https://bugs.gentoo.org/831487 but dropped system-ffmpeg.

They have their reasons, I have mine, that's why my ebuilds are still here and not merged upstream. See also https://github.com/PF4Public/gentoo-overlay/issues/262

> Again, following main tree

Oh right. I removed that line for my convenience :) I need it to continue no matter what, this way I can inspect logs and see all the errors all lat once without me needing to restart the build upon each error.

> As above, unclear.

Those were added specifically for cfi and I keep them

> Thanks, it's #142. Should add a comment.

Yes

> Uses all hyper-threads. See https://clang.llvm.org/docs/ThinLTO.html#controlling-backend-parallelism.

What if I use `MAKEOPTS=-j1`? You see, it's a bit tricky here :)

> The only possibility is that it has effect, see [rui314/mold#336](https://github.com/rui314/mold/issues/336). I have searched for the reason for quite a long time before...

If it works for you, then that's OK, I guess

PF4Public commented:
> > > `if use x86; then`
> > 
> > 
> > [Excuse me](https://github.com/PF4Public/ungoogled-chromium-bin/releases/tag/132.0.6834.159#:~:text=i686.tar.bz2,last%20week)?
> 
> What do you mean? It now just do `filter-flags "-g*"` all the time. x86 support is well retained.

It'll prevent debug builds. x86 cannot build it anyway, but I'd like to keep it for others.

PF4Public commented:
> Should be OK with NODIE check.

NODIE is set in GitHub runner, therefore it will fail there, that's what I was trying to avoid.

> What's more interesting is that mold supports `-Wl,--threads=N` and `-Wl,--thinlto-jobs=N` at the same time. So as it's hard to figure it out, let the users set it themselves.

But the default behaviour itself should be more or less predictable, that's why I find following MAKEOPTS as a most suitable common ground.

blackteahamburger commented:
> > Should be OK with NODIE check.
> 
> NODIE is set in GitHub runner, therefore it will fail there, that's what I was trying to avoid.

✅
> > What's more interesting is that mold supports `-Wl,--threads=N` and `-Wl,--thinlto-jobs=N` at the same time. So as it's hard to figure it out, let the users set it themselves.
> 
> But the default behaviour itself should be more or less predictable, that's why I find following MAKEOPTS as the most suitable common ground.

Then you mean `-Wl,--thinlto-jobs=$(makeopts_jobs)` should be explicitly set everywhere because "By default, the ThinLTO link step will launch as many threads in parallel as there are cores." However, there must be a reason why the default is this way (and may play nice with MAKEOPTS). You could consider asking upstream, but I don't think it's our responsibility to set it explicitly. Also, it blocks any user settings.

PF4Public commented:
> You could consider asking upstream

I'm not asking anyone, I decide for myself!

blackteahamburger commented:
> > You could consider asking upstream
> 
> I'm not asking anyone, I decide for myself!

You may have misunderstood what I meant because I didn't ask you to actually ask. 

After testing, I unfortunately noticed that it has no effect when thinlto-jobs is higher than MAKEOPTS!

PF4Public commented:
> You may have misunderstood what I meant because I didn't ask you to actually ask.

Sorry

> After testing, I unfortunately noticed that it has no effect when thinlto-jobs is higher than MAKEOPTS!

What does it mean for chromium-ebuilds? How should we handle that?

blackteahamburger commented:
> > After testing, I unfortunately noticed that it has no effect when thinlto-jobs is higher than MAKEOPTS!
> 
> What does it mean for chromium-ebuilds? How should we handle that?

Removing `-Wl,--thinlto-jobs=$(makeopts_jobs)` is just fine.

PF4Public commented:
> Removing `-Wl,--thinlto-jobs=$(makeopts_jobs)` is just fine.

This would break the expectation associated with MAKEOPTS, I'm not eager to do so.
